src/python/foglamp/core/scheduler/scheduler.py

```diff
@@ -948,208 +581,324 @@ def _schedule_first_task(self, schedule, current_time):
+                res = conn.update_tbl("tasks", update_payload)
+        except Exception:
+            self._logger.exception('Update failed: %s', update_payload)
+            raise
+
+    async def _read_config(self):
+        """Reads configuration"""
+        default_config = {
+            "max_running_tasks": {
+                "description": "The maximum number of tasks that can be running at any given time",
+                "type": "integer",
+                "default": str(self._DEFAULT_MAX_RUNNING_TASKS)
+            },
+            "max_completed_task_age_days": {
+                "description": "The maximum age, in days (based on the start time), for a rows "
+                               "in the tasks table that do not have a status of 'running'",
+                "type": "integer",
+                "default": str(self._DEFAULT_MAX_COMPLETED_TASK_AGE_DAYS)
+            },
+        }
+
+        await configuration_manager.create_category('SCHEDULER', default_config,
+                                                    'Scheduler configuration')
+
+        config = await configuration_manager.get_category_all_items('SCHEDULER')
+
+        self._max_running_tasks = int(config['max_running_tasks']['value'])
+        self._max_completed_task_age = datetime.timedelta(
+            seconds=int(config['max_completed_task_age_days']['value']) * self._DAY_SECONDS)
+
+    async def start(self):
+        """Starts the _scheduler
+
+        When this method returns, an asyncio task is
+        scheduled that starts tasks and monitors their subprocesses. This class
+        does not use threads (tasks run as subprocesses).
+
+        Raises:
+            NotReadyError: Scheduler was stopped
         """
-        schedule_execution = self._schedule_executions[schedule.id]
-        advance_seconds = schedule.repeat_seconds
+        if self._paused or self._schedule_executions is None:
+            raise NotReadyError("The _scheduler was stopped and can not be restarted")
 
-        if self._paused or advance_seconds is None:
-            schedule_execution.next_start_time = None
-            self._logger.info(
-                "Tasks will no longer execute for schedule '%s'", schedule.name)
+        if self._ready:
             return
 
-        now = time.time()
+        if self._start_time:
+            raise NotReadyError("The _scheduler is starting")
 
-        if (schedule.exclusive and schedule_execution.next_start_time and
-                now < schedule_execution.next_start_time):
-            # The task was started manually
-            # Or the schedule was modified after the task started (AVOID_ALTER_NEXT_START)
+        self._logger.info("Starting")
+
+        self._start_time = self.current_time if self.current_time else time.time()
+
+        # make sure that it go forward only when storage service is ready
+        storage_service = None
+        attempts_left = 3
```

Comment:
can we make it 
``` 
max_wait = 30
attempts_left = max_wait / 5
asyncio.sleep(5)

``` 
```
signal.setitimer(signal.ITIMER_REAL, max_wait)

```
to save 5 seconds if received in any of the attempt; cost will be minimal as its in-memory object check

Reply:
The original plan I had was to make things that depended on the storage service being there to be triggered off the registration coming form the storage service rather than have some form of polling or busy wait going on. Hence the words int he spec about callbacks that are triggered by the registration. However we didn't implement that and now os no the time to do that. It is highly questionable about having limits on the number of sleeps we do waiting for the storage service, if it doesn't come up then there is little we can do to continue. Rather than change this at this stage I think we should continue using the mechanism you have implemented for now and allow us to complete this sprint.